dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,663 @@
+ * parallel. Writes to the same document will be executed sequentially.
+ *
+ * @class
+ */
+export class BulkWriter {
+  // The maximum number of writes that can be in a single batch.
+  private maxBatchSize = DEFAULT_BULK_WRITER_MAX_BATCH_SIZE;
+
+  /**
+   * A queue of batches to be written.
+   */
+  private batchQueue: BulkCommitBatch[] = [];
+
+  /**
+   * A set of all document references that are currently in flight.
+   */
+  private refsInFlight = new Set<DocumentReference>();
+
+  /**
+   * Counter used to generate ids for batches.
+   */
+  private currentId = 0;
+
+  /**
+   * Whether this BulkWriter instance is closed. Once closed, it cannot be
+   * opened again.
+   */
+  private closed = false;
+
+  /**
+   * The maximum number of operations per second as determined by the 500/50/5
+   * rule. This value gradually increases with time.
+   */
+  private maxOpsPerSecond = STARTING_MAXIMUM_OPS_PER_SECOND;
+
+  constructor(
+    private readonly firestore: Firestore,
+    private readonly options?: BulkWriterOptions
+  ) {
+    if (options && !options.disableThrottling) {
+      this.rampMaxOps();
+    }
+  }
+
+  /**
+   * Create a document with the provided data. This single operation will fail
+   * if a document exists at its location.
+   *
+   * @param {DocumentReference} documentRef A reference to the document to be
+   * created.
+   * @param {T} data The object to serialize as the document.
+   * @returns {Promise<WriteResult>} A promise that resolves with the result of
+   * the write. Throws an error if the write fails.
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   * let documentRef = firestore.collection('col').doc();
+   *
+   * bulkWriter
+   *  .create(documentRef, {foo: 'bar'})
+   *  .then(result => {
+   *    console.log('Successfully executed write at: ', result);
+   *  })
+   *  .catch(err => {
+   *    console.log('Write failed with: ', err);
+   *  });
+   * });
+   */
+  create(
+    documentRef: DocumentReference,
+    data: DocumentData
+  ): Promise<WriteResult> {
+    this.verifyNotClosed();
+    const bulkCommitBatch = this.getEligibleBatch(documentRef);
+    return bulkCommitBatch.create(documentRef, data);
+  }
+
+  /**
+   * Delete a document from the database.
+   *
+   * @param {DocumentReference} documentRef A reference to the document to be
+   * deleted.
+   * @param {Precondition=} precondition A precondition to enforce for this
+   * delete.
+   * @param {Timestamp=} precondition.lastUpdateTime If set, enforces that the
+   * document was last updated at lastUpdateTime. Fails the batch if the
+   * document doesn't exist or was last updated at a different time.
+   * @returns {Promise<WriteResult>} A promise that resolves with the result of
+   * the write. Throws an error if the write fails.
+   *
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   * let documentRef = firestore.collection('col').doc();
```

Comment:
Since this example is for deleting a document, consider adding a document name argument to the `doc()` method call to demonstrate getting an *existing* document to delete.  As it is, calling `doc()` without an argument returns a document with "an automatically-generated unique ID" (i.e. it does not exist).